firedrake/slate/slac/compiler.py

```diff
@@ -192,6 +192,10 @@ def generate_loopy_kernel(slate_expr, compiler_parameters=None):
     get_index = lambda n: orig_coeffs.index(new_coeffs[n]) if new_coeffs[n] in orig_coeffs else n
```

Comment:
I think I have to understand these lines. `slate_expr.coefficients()` includes `BlockFunction` objects, but is it expected? Related is `get_index`; is it assuming something about how we collect coefficients?

Reply:
Yes it's expected that `BlockFunctions` are included here, I have just pushed a fix to account for that in the collection of the indices here. 
And yes we assume that the coefficients in `new_coeffs` or `orig_coeff` are collected in a specific order, which is the one they are appearing in in the corresponding expression.

Reply:
Quick question: is there a way to collect `BlockFunction` objects given a slate expression?

Reply:
This is untested but sth like this should work:
```
is_blockf = lambda s: True if isinstance(s, BlockFunction) else False
blockfuncs = filter(is_blockf, slate_expr.coefficients())
```

Reply:
You can see something similar in my fix here https://github.com/firedrakeproject/firedrake/pull/2446/commits/35f78e03867f388b700e84c1681ceee7c05e51ee

Reply:
Thanks. I'm just trying to see if we can simplify the implementation.

Reply:
So I think the way `get_index` currently works is somewhat fragile; I think we should assume that people can do all sorts of optimisation which can reorder coefficients. Also, I still think `collect_coefficients()` should loop over `coeff_map` to make sure that the order of kernerl arguments and that of coefficients (collected for PyOP2) would be the same.

Do you think the following works, or am I missing something?

- Add `raw_coeffcients()` method:
```
def raw_coefficients(self):
    return self.coefficients()  # change this for `BlockedAssembledVector`
```
- `slate/slac/compiler.py`:
```
...
orig_coeffs = orig_expr.coefficients()
new_coeffs = slate_expr.raw_coefficients()
map_new_to_orig = [orig_coeffs.index(c) for c in new_coeffs]
coeff_map = ((map_new_to_orig(n), split_map) for (n, split_map) in slate_expr.coeff_map)
...
```

- `slate/slate.py`:
Change `coeff_map` so that the target index would be `m = self.raw_coefficients().index(some_raw_coeff)` instead of incrementing `m`.
Sort collected indices.

- `slate/slac/kernel_builder.py`:
```
    def collect_coefficients(self):
        coeffs = self.expression.raw_coefficients()
        coeff_dict = OrderedDict()
        for i, (c, split_map) in enumerate(self.expression.coeff_map):
            coeff = coeffs[c]
            if type(coeff.ufl_element()) == MixedElement:
                mixed = OrderedDict() if coeff not in coeff_dict else coeff_dict[coeff]
                for j in split_map:
                    coeff_ = coeff.split()[j]
                    mixed[coeff_] = f"w_{i}_{j}", self.extent(coeff_)
                coeff_dict.update({coeff: mixed})
            else:
                coeff_dict[coeff] = f"w_{i}", self.extent(coeff)
        return coeff_dict
```